File: Data Structures & Algorithms/valid-binary-search-tree/submission-10.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        prev = -1001

        def inorder(node):
            nonlocal prev
            if not node:
                return True

            if not inorder(node.left):
                return False

            if node.val <= prev:
                return False

            prev = node.val

            if not inorder(node.right):
                return False

            return True


        return inorder(root)

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-10.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    lowestCommonAncestor(
        root: TreeNode | null,
        p: TreeNode | null,
        q: TreeNode | null,
    ) {
        let a = root

        while (true) {
            if (Math.min(p.val, q.val) > a.val) {
                a = a.right
            } else if (Math.max(p.val, q.val) < a.val) {
                a = a.left
            } else {
                return a
            }
        }
    }
}
